File: python/problems/leetcode/0114_flatten_binary_tree_to_linked_list/solution.py
"""LeetCode 114: Flatten Binary Tree to Linked List."""

from __future__ import annotations

from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        cur = root
        while cur is not None:
            if cur.left is not None:
                rightmost = cur.left
                while rightmost.right is not None:
                    rightmost = rightmost.right
                rightmost.right = cur.right
                cur.right = cur.left
                cur.left = None
            cur = cur.right


def to_right_chain(root: Optional[TreeNode]) -> list[int]:
    out: list[int] = []
    while root is not None:
        assert root.left is None
        out.append(root.val)
        root = root.right
    return out


if __name__ == "__main__":
    root = TreeNode(
        1,
        TreeNode(2, TreeNode(3), TreeNode(4)),
        TreeNode(5, None, TreeNode(6)),
    )
    Solution().flatten(root)
    assert to_right_chain(root) == [1, 2, 3, 4, 5, 6]
    Solution().flatten(None)
    one = TreeNode(0)
    Solution().flatten(one)
    assert to_right_chain(one) == [0]
    print("0114 OK")

File: cpp/problems/leetcode/0114_flatten_binary_tree_to_linked_list/solution.cpp
// LeetCode 114
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    void flatten(TreeNode* root) {
        TreeNode* cur = root;
        while (cur) {
            if (cur->left) {
                TreeNode* rm = cur->left;
                while (rm->right) rm = rm->right;
                rm->right = cur->right;
                cur->right = cur->left;
                cur->left = nullptr;
            }
            cur = cur->right;
        }
    }
};

int main() {
    TreeNode n3(3), n4(4), n2(2, &n3, &n4), n6(6), n5(5, nullptr, &n6);
    TreeNode root(1, &n2, &n5);
    Solution().flatten(&root);
    vector<int> out;
    for (TreeNode* p = &root; p; p = p->right) {
        assert(p->left == nullptr);
        out.push_back(p->val);
    }
    assert((out == vector<int>{1, 2, 3, 4, 5, 6}));
    Solution().flatten(nullptr);
    TreeNode one(0);
    Solution().flatten(&one);
    assert(one.left == nullptr && one.right == nullptr && one.val == 0);
    cout << "0114 OK" << endl;
    return 0;
}
